src/github.com/couchbase/sync_gateway/db/channel_cache.go

```diff
@@ -196,6 +206,46 @@ func (c *channelCache) _appendChange(change *LogEntry) {
 	c.logs = append(log, change)
 }
 
+// Insert out-of-sequence entry into the cache.  If the docId is already present in a later
+// sequence, we skip the insert.  If the docId is already present in an earlier sequence,
+// we remove the earlier sequence.
+func insertChange(log *LogEntries, change *LogEntry) {
+
+	end := len(*log) - 1
+	insertAfterIndex := -1
+	for i := end; i >= 0; i-- {
+		currLog := (*log)[i]
+		if insertAfterIndex == -1 && currLog.Sequence < change.Sequence {
+			insertAfterIndex = i
+		}
+		if currLog.DocID == change.DocID {
+			if currLog.Sequence >= change.Sequence {
+				// we've already cached a later revision of this document, can ignore update
+				return
+			} else {
+				// found existing prior to insert position
+				if i == insertAfterIndex {
+					// The sequence is adjacent to another with the same docId - replace it
+					// instead of inserting
+					(*log)[i] = change
+					return
+				} else {
+					// Shift and insert to remove the old entry and add the new one
+					copy((*log)[i:insertAfterIndex], (*log)[i+1:insertAfterIndex+1])
+					(*log)[insertAfterIndex] = change
+					return
+				}
+			}
+		}
+	}
+	// We didn't find a match for DocID, so standard insert.
+	*log = append(*log, nil)
+	copy((*log)[insertAfterIndex+2:], (*log)[insertAfterIndex+1:])
+	(*log)[insertAfterIndex+1] = change
+
```

Comment:
Lines 243 and 244 are hard to understand.  Maybe turning `insertAfterIndex+2` and `insertAfterIndex+1` into clearly named variables would help, and/or adding an explanation of what is happening on these lines.